pkg/endpoint/bpf.go

```diff
@@ -410,6 +410,38 @@ func (e *Endpoint) regenerateBPF(regenContext *regenerationContext) (revnum uint
 		return 0, err
 	}
 
+	// Avoid BPF program compilation and installation if the endpoint or node
+	// configuration hasn't changed. Hashing the endpoint configuration requires
+	// getting the security identity, which takes out a read lock on the Endpoint.
+	// Make sure to calculate the endpoint hash outside of a locked context.
+	datapathRegenCtxt.bpfHeaderfilesHash, err = e.owner.Orchestrator().EndpointHash(e)
+	if err != nil {
+		return 0, fmt.Errorf("hash endpoint configuration: %w", err)
+	}
+
+	if datapathRegenCtxt.bpfHeaderfilesHash != e.bpfHeaderfileHash {
```

Comment:
Just checking, we don't need to hold an endpoint lock for accessing `e.bpfHeaderfileHash`?

Reply:
Both accesses happen within the scope of `regenerateBPF`, under the compilation read lock, but honestly I can't answer your question with certainty. One thing that may work in our favor is that `EndpointHash()` takes out the Endpoint's write lock, so if there are any policy updates going on, this will block until that's done.

I think the worst case is that the endpoint gets regenerated when it shouldn't. I hope to give this part of the loader a thorough rework to do away with the rampant complexity, it's not really possible to reason about in its current state IMO.